<a href="https://colab.research.google.com/github/manfredialdo/iad-tpi01/blob/main/Manfredi_Aldo_10_Entrega2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## . Datos de identificación

- **Apellido y nombre:** manfredi aldo dario
- **Comisión:** 10
- **Dataset asignado:** hotel booking TPI grupo J.csv
- **Entrega:** Trabajo practico integrador primera entrega
- **Fecha:** Domingo 6 de septiembre de 2026, 23:59 hs.
- **Repositorio:** https://github.com/manfredialdo/iad-tpi01
- **Archivo:** Manfredi_Aldo_10_Entrega2.ipynb

# 🧩 TPI · Entrega 2 · Limpieza, preparación y transformación de datos (OBLIGATORIA)

En esta entrega se debe aplicar al dataset asignado los procedimientos trabajados durante la Unidad N° 2. El objetivo es diagnosticar la calidad de los datos, justificar decisiones de limpieza y construir las transformaciones necesarias para preparar el dataset para las próximas etapas del análisis.

* **Caso de estudio:** Reservas hoteleras
* **Variable objetivo:** `is_canceled`
* **Lenguaje a elección:** Python o R (debes elegir y utilizar un solo lenguaje).
* **⏱️ Tiempo estimado:** 2 pomodoros de 20 minutos (continuidad de la Semana 4).

---

## 💻 Datos del Estudiante e Identificación

> **Atención:** Completa tus datos personales antes de comenzar con la resolución del notebook.

* **Apellido y Nombre:** manfredi aldo dario
* **Comisión:** 10
* **Número de Entrega:** Entrega 2
* **Dataset Asignado:**  dataset J https://tup.sied.utn.edu.ar/pluginfile.php/32860/course/section/4282/hotel%20booking%20TPI%20grupo%20J.csv
* **Problema de Análisis:**
* **Variable Objetivo:** `is_canceled`

---

## 📝 Requisitos Técnicos de Carga

1. **Notebook independiente:** Debe poder ejecutarse desde el inicio de manera fluida sin depender de archivos de entregas anteriores.
2. **Carga reproducible:** Importación de bibliotecas necesarias y carga directa del dataset original.
3. **Copia de trabajo:** Conservar el dataset original intacto en una variable y crear una copia explícita sobre la cual se aplicarán las modificaciones.

---

## 📋 Contenido Obligatorio de la Entrega

1. **Diagnóstico de calidad:** Revisión de valores faltantes, duplicados, tipos de datos, categorías inconsistentes, valores inválidos y posibles valores atípicos, manteniendo la bitácora correspondiente.
2. **Transformación de variables:** Aplicación justificada de conversiones temporales, codificación, discretización, normalización o estandarización cuando resulte pertinente.
3. **Variables derivadas:** Creación de nuevas columnas útiles para el problema (ej. duración total de la estadía, cantidad total de huéspedes, importe estimado, reserva familiar o segmentos de anticipación).
4. **Bitácora del proceso:** Registro estructurado con las variables transformadas, el problema detectado, la decisión aplicada y su justificación.
5. **Dataset preparado:** Exposición del estado final del conjunto de datos y explicación de cómo las decisiones tomadas lo dejan listo para análisis posteriores.

> 📌 **Criterio metodológico:** No se espera que apliques todas las transformaciones posibles ni que elimines automáticamente cada valor inusual. Cada procedimiento debe responder a un problema o necesidad concreta y estar acompañado por una justificación.

---

## 🔎 Variables Sugeridas para Revisión

| Variable(s) | Aspecto a revisar / Transformación |
| :--- | :--- |
| `children`, `adults`, `babies` | Cantidad total de huéspedes e Inconsistencias. |
| `stays_in_weekend_nights`, `stays_in_week_nights` | Duración total de la estadía. |
| `adr` | Valores faltantes, negativos, ceros o atípicos. |
| `lead_time` | Anticipación de la reserva y valores elevados. |
| `country`, `agent`, `company` | Presencia e interpretación de valores ausentes. |
| `hotel`, `market_segment`, `distribution_channel`, `deposit_type`, `customer_type` | Revisión y posible codificación de variables categóricas. |
| **Variables de fecha de llegada** | Formato de fecha y construcción de variables temporales. |

---

## ✅ Lista de Comprobación (Checklist de Entrega)

- [ ] ¿El notebook identifica al estudiante, la comisión y la entrega?
- [ ] ¿Puede ejecutarse de principio a fin sin depender de otro notebook?
- [ ] ¿Conserva el dataset original intacto y aplica los cambios sobre una copia?
- [ ] ¿Revisa los principales problemas de calidad de los datos?
- [ ] ¿Justifica formalmente cada decisión de limpieza y transformación?
- [ ] ¿Las variables derivadas están correctamente construidas y calculadas?
- [ ] ¿Combina explicaciones en texto con bloques de código y sus resultados?
- [ ] ¿Explica detalladamente cómo quedó preparado el dataset final?

---

## ⚠️ Condiciones y Pautas de Entrega

* **Estado:** Obligatoria
* **Habilitación:** Lunes 31 de agosto de 2026, 00:00 hs.
* **Fecha límite de entrega:** Domingo 6 de septiembre de 2026, 23:59 hs.
* **Formato:** Único archivo `.ipynb`.
* **Tamaño máximo:** 20 MB.
* **Calificación mínima de aprobación:** 6 / 10 puntos.
* **Intentos permitidos:** Hasta 2 intentos (la reaparición será habilitada por el tutor cuando corresponda).
* **Estructura de nombre del archivo:** `Apellido_Nombre_Comision_Entrega2.ipynb`

> 💡 **Recomendación:** Conserva también una versión editable acumulativa (notebook maestro) donde integres progresivamente las distintas etapas del TPI para la entrega final.

carga del dataset


In [1]:
import pandas as pd
# Carga del archivo CSV
url = "https://raw.githubusercontent.com/manfredialdo/iad-tpi01/refs/heads/main/hotel%20booking%20TPI%20grupo%20J.csv"
df = pd.read_csv(url)

### 0- Preparativos

Primero vamos a hacer una copia del dataset original para dejarlo como respaldo. Es una buena practica cuando se está modificando los datos. En caso de cometer algún error o inconveniente, se puede volver al dataset original y recuperar estos datos.

In [2]:
# Conserva el dataset original intacto y aplica los cambios sobre una copia?
df_limpio = df.copy()

In [3]:
print(df_limpio.shape)

(25000, 32)


Posee 25000 filas y 32 columnas

In [6]:
# tipos de variables
df_limpio.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25000 entries, 0 to 24999
Data columns (total 32 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   booking_id                      25000 non-null  object 
 1   hotel                           25000 non-null  object 
 2   is_canceled                     25000 non-null  int64  
 3   lead_time                       25000 non-null  int64  
 4   arrival_date_year               25000 non-null  int64  
 5   arrival_date_month              25000 non-null  object 
 6   arrival_date_week_number        25000 non-null  int64  
 7   arrival_date_day_of_month       25000 non-null  int64  
 8   arrival_date                    25000 non-null  object 
 9   stays_in_weekend_nights         25000 non-null  int64  
 10  stays_in_week_nights            25000 non-null  int64  
 11  adults                          25000 non-null  int64  
 12  children                        

In [8]:
# resumen estadistico
df_limpio.describe()

,is_canceled,lead_time,arrival_date_year,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,is_repeated_guest,previous_cancellations,previous_bookings_not_canceled,booking_changes,agent,company,days_in_waiting_list,adr,required_car_parking_spaces,total_of_special_requests
count,25000.000000,25000.000000,25000.000000,25000.000000,25000.000000,25000.000000,25000.000000,25000.000000,25000.000000,25000.000000,25000.000000,25000.000000,25000.000000,25000.000000,21521.000000,1440.000000,25000.00000,25000.000000,25000.000000,25000.000000
mean,0.370360,104.223240,2024.156640,26.689680,15.805480,0.920440,2.499000,1.854440,0.103080,0.008040,0.033240,0.090360,0.143240,0.229080,86.426746,187.325694,2.27540,101.681678,0.062960,0.567880
std,0.482911,107.098048,0.707421,13.409251,8.748617,0.998193,1.903613,0.571371,0.393218,0.091519,0.179266,0.873559,1.613576,0.644033,111.048259,128.592996,17.07986,47.969113,0.243718,0.791693
min,0.000000,0.000000,2023.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,9.000000,0.00000,0.000000,0.000000,0.000000
25%,0.000000,18.000000,2024.000000,16.000000,8.000000,0.000000,1.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,9.000000,62.000000,0.00000,69.027500,0.000000,0.000000
50%,0.000000,69.000000,2024.000000,27.000000,16.000000,1.000000,2.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,14.000000,186.000000,0.00000,94.500000,0.000000,0.000000
75%,1.000000,160.000000,2025.000000,37.000000,23.000000,2.000000,3.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,229.000000,268.000000,0.00000,126.000000,0.000000,1.000000
max,1.000000,629.000000,2025.000000,52.000000,31.000000,18.000000,42.000000,40.000000,3.000000,2.000000,1.000000,26.000000,72.000000,18.000000,535.000000,541.000000,391.00000,387.000000,3.000000,5.000000


In [ ]:
df_limpio.describe(include=['object','category'])

 Conclusiones parciales del Análisis Exploratorio de Datos

---

1. Cancelaciones y Comportamiento de Reserva
* **Alta tasa de cancelación (37%):** Casi 4 de cada 10 reservas terminan cancelándose (`is_canceled` medio = `0.37`).
* **Antelación de reserva (*Lead Time*) elevado:** La media es de **104 días** (con un máximo extremo de 629 días). El 50% de los clientes reserva con más de 69 días de anticipación.
* **Depósitos y políticas flexibilizadas:** El **87.5%** de las reservas no incluyen depósito previo (`No Deposit` = `21,880`). Existe una correlación directa habitual entre la falta de depósito, un alto *lead time* y la elevada tasa de cancelación.

---

2. Perfil de la Demanda y Canales de Venta
* **Predominio del turismo urbano:** El **66.5%** de las reservas corresponden a la categoría *City Hotel* (`16,614`).
* **Dominio absoluto de intermediarios digitales:** El canal `TA/TO` (Agencias y Operadores de Turismo) representa el **81.5%** de las reservas (`20,384`), siendo el segmento `Online TA` el principal motor (`11,687`).
* **Mercado local fuerte:** El país de origen predominante es Portugal (`PRT`), representando el **40.8%** de los huéspedes con país registrado (`10,159`).
* **Mes de mayor demanda:** Agosto es la temporada alta con el mayor volumen de registros (`2,905`).

---

3. Comportamiento del Huésped y Estancias
* **Tipo de cliente y composición:**
  * El **75.2%** son clientes de tipo *Transient* (turistas o viajeros individuales sin contrato corporativo).
  * La ocupación promedio es de **2 adultos** por reserva (`mean` = `1.85`). La presencia de niños o bebés es residual (`0.10` y `0.008` de media).
* **Duración de la estancia:** La media de noches en días de semana es de **2.5 noches**, mientras que los fines de semana promedian **0.9 noches** (~3 a 4 noches totales por reserva).
* **Poca fidelización:** Solo el **3.3%** son clientes repetidos (`is_repeated_guest` = `0.033`).

---

4. Tarifas y Preferencias Operativas
* **Tarifa Promedio Diaria (*ADR*):** El precio medio por habitación/noche es de **$101.68** (con el 50% de las reservas entre $69 y $126).
* **Régimen alimenticio y habitación:**
  * El **77.2%** elige alojamiento y desayuno (`BB` - *Bed & Breakfast*).
  * El tipo de habitación **A** es la más reservada (`18,038`) y asignada (`15,545`). Sin embargo, hay discrepancias entre habitación reservada y asignada, indicando situaciones de *upgrade* o reasignación por disponibilidad.
* **Baja demanda de estacionamiento:** La gran mayoría de los clientes no solicita cochera (media de `0.06` espacios por reserva).

---

5. Calidad de los Datos (*Data Quality*)
* **Datos faltantes:** La variable `country` tiene **117 valores nulos** (`25,000` - `24,883`).
* **Variables con alto vaciado / nulos parciales:** `agent` y `company` presentan un conteo menor a `25,000` (`21,521` y `1,440` respectivamente), lo que indica que muchas reservas son directas o no corporativas.
* **Valores atípicos (*Outliers*):**
  * `adults`: Muestra un valor máximo de **40 adultos** en una sola reserva (posible error de carga o reserva de grupo en un solo ID).
  * `stays_in_week_nights`: Máximo de **42 noches** de semana.
  * `days_in_waiting_list`: Máximo de **387 días** en lista de espera.

También, crearemos un documento que llamaremos "Bitácora" para dejar constancia del problema detectado, las variables afectadas, la decisión tomada y la justificación de dicha decisión.

In [5]:
registros_bitacora = []

def registrar(problema, variable, decision, justificacion):
    registros_bitacora.append({
        "problema_detectado": problema,
        "variable_afectada": variable,
        "decision_tomada": decision,
        "justificacion": justificacion
    })

In [4]:
df_limpio.columns

Index(['booking_id', 'hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'arrival_date', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests'],
      dtype='object')

In [10]:
df_limpio.duplicated().sum()

np.int64(0)

observamos que no existen valores duplicados. Por lo tanto, no necesitamos realizar ninguna acción en este paso y contiuamos con el resto de la limpieza de datos.

In [11]:
df_limpio.isna().sum()

,0
booking_id,0
hotel,0
is_canceled,0
lead_time,0
arrival_date_year,0
arrival_date_month,0
arrival_date_week_number,0
arrival_date_day_of_month,0
arrival_date,0
stays_in_weekend_nights,0


🧩 Análisis de Valores Nulos (`isna().sum()`)

El diagnóstico de datos faltantes muestra un conjunto de datos con **excelente nivel de integridad**, donde **30 de las 33 columnas no presentan ningún valor nulo**.

* **`country` (117 nulos - 0.47%):** Representa una cantidad residual de reservas sin país registrado.
* **`agent` (3,479 nulos - 13.92%):** Indica reservas realizadas directamente por el cliente sin intermediación de una agencia.
* **`company` (23,560 nulos - 94.24%):** Refleja que la gran mayoría de los huéspedes son viajeros particulares y no corporativos.

> **Estrategia de Limpieza:** Se recomienda imputar `agent` y `company` con el valor `0` (indicando ausencia de agente/empresa) e imputar `country` con la moda (`PRT`) o la categoría `'Unknown'`.